mockgcp/mockbigtable/instance.go

```diff
@@ -0,0 +1,250 @@
+)
+
+type instanceAdminServer struct {
+	*MockService
+	pb.UnimplementedBigtableInstanceAdminServer
+}
+
+func (s *instanceAdminServer) GetInstance(ctx context.Context, req *pb.GetInstanceRequest) (*pb.Instance, error) {
+	name, err := s.parseInstanceName(req.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	fqn := name.String()
+
+	obj := &pb.Instance{}
+	if err := s.storage.Get(ctx, fqn, obj); err != nil {
+		if status.Code(err) == codes.NotFound {
+			return nil, status.Errorf(codes.NotFound, "Instance %s not found.", name.InstanceName)
+		}
+		return nil, err
+	}
+
+	return obj, nil
+}
+
+func (s *instanceAdminServer) ListInstances(ctx context.Context, req *pb.ListInstancesRequest) (*pb.ListInstancesResponse, error) {
+	projectName, err := projects.ParseProjectName(req.GetParent())
+	if err != nil {
+		return nil, err
+	}
+	project, err := s.Projects.GetProject(projectName)
+	if err != nil {
+		return nil, err
+	}
+
+	response := &pb.ListInstancesResponse{}
+	findKind := (&pb.Instance{}).ProtoReflect().Descriptor()
+	if err := s.storage.List(ctx, findKind, storage.ListOptions{
+		Prefix: "projects/" + project.ID + "/",
+	}, func(obj proto.Message) error {
+		instance := obj.(*pb.Instance)
+		response.Instances = append(response.Instances, instance)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return response, nil
+}
+
+func (s *instanceAdminServer) CreateInstance(ctx context.Context, req *pb.CreateInstanceRequest) (*longrunning.Operation, error) {
+	reqName := req.GetParent() + "/instances/" + req.GetInstanceId()
+	name, err := s.parseInstanceName(reqName)
+	if err != nil {
+		return nil, err
+	}
+
+	// "Clean up" the original request; this syncs the metadata OriginalRequest field
+	for _, reqCluster := range req.GetClusters() {
+		if proto.Equal(reqCluster.EncryptionConfig, &pb.Cluster_EncryptionConfig{}) {
+			reqCluster.EncryptionConfig = nil
+		}
+		if reqCluster.DefaultStorageType == pb.StorageType_STORAGE_TYPE_UNSPECIFIED {
+			reqCluster.DefaultStorageType = pb.StorageType_SSD
+		}
+		reqCluster.Name = ""
+	}
+
+	originalRequest := proto.Clone(req).(*pb.CreateInstanceRequest)
+
+	now := time.Now()
+	instanceFQN := name.String()
+
+	obj := proto.Clone(req.Instance).(*pb.Instance)
+	obj.Name = instanceFQN
+
+	obj.State = pb.Instance_READY
+	obj.CreateTime = timestamppb.New(now)
+
+	if err := s.populateDefaultsForInstance(obj); err != nil {
+		return nil, err
+	}
+
+	if err := s.storage.Create(ctx, instanceFQN, obj); err != nil {
+		return nil, err
+	}
+
+	// If this was production, we'd probably want a transaction etc
+	for clusterID, cluster := range req.GetClusters() {
+		clusterFQN := instanceFQN + "/clusters/" + clusterID
+		obj := proto.Clone(cluster).(*pb.Cluster)
+		obj.Name = clusterFQN
+		if err := s.populateDefaultsForCluster(obj); err != nil {
+			return nil, err
+		}
+		if err := s.storage.Create(ctx, clusterFQN, obj); err != nil {
+			return nil, err
+		}
+	}
+
+	zone := "us-west1-b" // TODO
```

Comment:
nit: "us-central1-a"? So the instance, cluster and table will be in the same zone.

Reply:
Good call.  It is odd that this is zonal TBH :-)  But we don't need to worry about it _too_ much for our mocks!